rfcs/text/0428-type-inheritance-and-forking.md

```diff
@@ -0,0 +1,952 @@
+
+As a consequence of the above definition, these questions arise:
+
+- How do we ensure that `Employee` instances can be used in place of `Person` instances in practice? (we refer to this as "subtyping")
+- Do multiple supertypes impose constraints on extending types?
+- How does having `additionalProperties` in existing schemas influence extended types?
+- How should the Block Protocol type system support extending types?
+- What type duplication strategy should the Block Protocol support?
+
+## Subtyping
+
+Compatibility between extended types is the ability to use the subtype in place of a supertype. [In academia](https://en.wikipedia.org/wiki/Subtyping#Coercions), this is `coercive subtyping` when it can happen implicitly.
+_Composition_ rather than inheritance allows us to have more guarantees about the relationship between subtypes and supertypes. If a subtype never modifies a supertype's inherited properties, we are sure that the supertype properties are left untouched, which eliminates the need for evaluating compatibility (i.e. no need for [`subsumption`](https://en.wikipedia.org/wiki/Subtyping#Subsumption), the concept of finding out whether or not a supertype is a supertype of a subtype or not).
+
+For example, an `Employee` instance looks as follows (simplified):
+
+```json
+{
+  "name": "Charles",
+  "age": 35,
+  "occupation": "Merchant"
+}
+```
+
+And the properties would have the following relation to `Employee`:
+
+```txt
+      name◄┐ (supertype)    (subtype)
+           ├───Person───────Employee
+       age◄┘                │
+                            │
+occupation◄─────────────────┘
+```
+
+We can visually see how selecting `Person` in the type hierarchy would provide `name` and `age` properties but exclude the `occupation` property.
+Assuming that we can project/select the properties of a type that are defined through the supertype, coercive subtyping is attainable for any subtype. This is a somewhat strong assumption to make, but it unlocks expressing how to extend types. The reason the selection/projection of properties is important for coercive subtyping will be explained in more detail in the coming sections, specifically in the [`additionalProperties` problem explanation section](#the-additionalproperties-problem).
+
+While it was mentioned that we don't allow overriding properties, we do allow it to a certain extent. When properties are still _compatible_ (using the definition of compatible given in the [versioning RFC](./0408-versioning-types.md#determining-type-compatibility)), properties can be overridden. For example, if `Person` had an optional property `name`, `Employee` could override it to be a required property. This is because the property is still compatible, but it is now more specific. This is also the reason why we can't allow overriding properties to be arrays or objects, as that would make the property incompatible.
```

Comment:
Maybe we should change the wording slightly:
```suggestion
While it was mentioned that we don't allow overriding properties, we do allow modifications to a certain extent. When properties are still _compatible_ (using the definition of compatible given in the [versioning RFC](./0408-versioning-types.md#determining-type-compatibility)), properties can be overridden. For example, if `Person` had an optional property `name`, `Employee` could override it to be a required property. This is because the property is still compatible, but it is now more specific. This is also the reason why we can't allow overriding properties to be arrays or objects, as that would make the property incompatible.
```